### ISSUE TITLE
CUDA+inference correctness issue with torchdynamo eager mode on TorchBench

### ISSUE THREAD

**Issue by xuzhao9** [pytorch/torchdynamo#165]
I did a TorchBench sweep run workflow on V100 and the result is available [here](https://github.com/pytorch/benchmark/actions/runs/2201634569). Or view the spreadsheet result [Meta internal only](https://docs.google.com/spreadsheets/d/1eQw4Y5x4ynOZx0Fgpm47ooWAortaTwM4SPxc1y1A-ew/edit#gid=966748872).
In this workflow, I have found the following correctness issues with TorchDynamo eager mode on CUDA inference tests:

- [x] dlrm (similarity: 0.0)
- [ ] fastNLP_Bert (similarity: 0.978)
- [ ] moco (similarity: 0.0)
- [x] pyhpc_turbulent_kinetic_energy (similarity: 0.0)
- [x] timm_efficientnet (similarity: 0.01)

The cosine similarity value is calculated comparing with the fp32 result.

**Comment by jansel** [comment]
@xuzhao9 these results contradict [the correctness checking in torchbench.py](https://github.com/facebookresearch/torchdynamo/blob/26cb2b01f389172c8bbce900a44009e698b818b5/torchbench.py#L1074).

When I run these models with `torchbench.py` they pass:
```
$ ./torchbench.py -k dlrm -k fastNLP_Bert -k moco -k timm_efficientnet -dcuda --backend=eager -n2
cuda eval  dlrm                               1.009x SAME
cuda eval  fastNLP_Bert                       0.989x SAME
cuda eval  moco                               1.001x SAME
cuda eval  timm_efficientnet                  1.019x SAME
speedup gmean=1.03x mean=1.03x
```

Note torchbench.py doesn't check `pyhpc_turbulent_kinetic_energy` because it is non-determinstic.  It fails `torch.allclose()` checks even if you just run it twice in eager mode and is not self-consistent.  But the other 4 pass.  Because of this I have excluded it from my results.

These tests are using [torch.allclose](https://github.com/facebookresearch/torchdynamo/blob/26cb2b01f389172c8bbce900a44009e698b818b5/torchdynamo/testing.py#L100), which I believe is a harder test to pass than cosine similarity.

Can you try to figure out what is different compared to your experimental setup?  Perhaps one of them have a bug.  I know torchbench.py does a lot of stuff to setup random state correctly.

**Comment by xuzhao9** [comment]
@jansel I can reproduce your result with the default same algorithm, which uses `torch.allclose`. However, after changing the parameter `cos_similarity=True`, I get the following result:

```
$ ./torchbench.py -k dlrm -dcuda --backend=eager -n2
cuda eval  dlrm                               Similarity score=0.0
INCORRECT - Variation in Eager runs itself
$ ./torchbench.py -k moco -dcuda --backend=eager -n2
cuda eval  moco                               Similarity score=1.0
Similarity score=0.0
INCORRECT - Variation in Eager runs itself
$ ./torchbench.py -k timm_efficientnet -dcuda --backend=eager -n2
cuda eval  timm_efficientnet                  Similarity score=0.988427460193634
INCORRECT - Variation in Eager runs itself
```

I can confirm fastNLP_Bert passes the correctness test

**Comment by jansel** [comment]
@xuzhao9 I think something is buggy in your setup.  

It is printing out `INCORRECT - Variation in Eager runs itself`, which is triggered [here](https://github.com/facebookresearch/torchdynamo/blob/6595fd900480ec01976a4abd8f3746c055515c05/torchbench.py#L1091).

You aren't even running TorchDynamo.  This printout is *before* TorchDynamo is run later down on [line 1098](https://github.com/facebookresearch/torchdynamo/blob/6595fd900480ec01976a4abd8f3746c055515c05/torchbench.py#L1098).  It is comparing eager with unmodified eager.

You are failing an A/A test, where absolutely nothing changed.  That test is meant to find bugs in benchmarks and measurement setups.

**Comment by xuzhao9** [comment]
@jansel Thanks! I talked to Wei and can confirm this is a bug in the torchbench repo. I have submitted https://github.com/pytorch/benchmark/pull/886, after landing it I will rerun the workflow

**Comment by xuzhao9** [comment]
With the fix, I can confirm that dlrm and timm_efficientnet are correct; moco and pyhpc_turbulent_kinetic_energy are unstable. However, for fastNLP_Bert, I get the following result with the `torchbench.py` tool:
```
$ python torchbench.py -k fastNLP_Bert -dcuda --backend=eager -n2
cuda eval  fastNLP_Bert                       1.094x p=0.02
STATS frames
  ('total', 14)
  ('ok', 14)
STATS graph_break
  ('Tensor.item', 3)
  ('numpy', 2)
  ('Tensor.numpy', 1)
  ('non-static call NumpyVariable() ConstDictVariable()', 1)
  ('call_function BuiltinVariable(list) [UserDefinedObjectVariable(chain)] {}', 1)
STATS inline_call
  ('Tensor.item', 2)
STATS stats
  ('calls_captured', 470)
  ('fusions_possible', 464)
  ('unique_graphs', 6)
STATS unimplemented
  ('no graph found', 3)
speedup gmean=1.09x mean=1.09x
```

@frank-wei can reproduce this as well, does this mean it fails the correctness test?

**Comment by jansel** [comment]
`p=0.02` is unrelated to correctness.  When correctness fails it will print out `INCORRECT` and not show any speedup.  So everything is running as expected in your log.

`p=` is a reference to a [p-value](https://en.wikipedia.org/wiki/P-value).  Torchbench.py it is doing a [t-test](https://en.wikipedia.org/wiki/Student%27s_t-test).  The purpose is to filter out noise from real speedups.

**Comment by xuzhao9** [comment]
@jansel Using this script, I can still reproduce the correctness problem of fastNLP_Bert:
``` 
import torch
import time
import gc
import copy
import sys
import numpy as np
from torchbenchmark import load_model_by_name
import argparse

DEVICE = "cuda"

def synchronize():
    if DEVICE == "cuda":
        torch.cuda.synchronize()

def same(a, b, cos_similarity=False, tol=1e-4):
    """Check correctness to see if a and b match"""
    if isinstance(a, (list, tuple, torch.nn.ParameterList, torch.Size)):
        assert isinstance(b, (list, tuple)), f"type mismatch {type(a)} {type(b)}"
        return len(a) == len(b) and all(
            same(ai, bi, cos_similarity, tol) for ai, bi in zip(a, b)
        )
    elif isinstance(a, dict):
        assert isinstance(b, dict)
        assert set(a.keys()) == set(
            b.keys()
        ), f"keys mismatch {set(a.keys())} == {set(b.keys())}"
        for k in a.keys():
            if not (same(a[k], b[k], cos_similarity, tol)):
                print("Accuracy failed for key name", k)
                return False
        return True
    elif isinstance(a, torch.Tensor):
        if a.is_sparse:
            assert b.is_sparse
            a = a.to_dense()
            b = b.to_dense()
        assert isinstance(b, torch.Tensor)
        if cos_similarity:
            # TRT will bring error loss larger than current threshold. Use cosine similarity as replacement
            a = a.flatten().to(torch.float32)
            b = b.flatten().to(torch.float32)
            res = torch.nn.functional.cosine_similarity(a, b, dim=0, eps=1e-6)
            print(f"Similarity score={res.cpu().numpy()}")
            return res >= 0.99
        else:
            return torch.allclose(a, b, atol=tol, rtol=tol)
    else:
        raise RuntimeError(f"unsupported type: {type(a).__name__}")

def run(model, example_inputs, times=1, dynamo=None):
    torch.manual_seed(1337)
    if dynamo:
        import torchdynamo
    synchronize()
    # Dont collect outputs to correctly measure timing
    results = []
    if dynamo:
        for inputs in example_inputs:
            batch_x, batch_y = inputs
            x = (batch_x["words"], )
            with torchdynamo.optimize("eager"):
                results.append(model(*x))
    else:
        for inputs in example_inputs:
            batch_x, batch_y = inputs
            x = (batch_x["words"], )
            results.append(model(*x))
    synchronize()
    return results

def correctness_experiment(model, example_inputs):
    repeat = 100
    timings = np.zeros((repeat, 2), np.float64)
    for rep in range(repeat):
        eager_outputs = run(model, example_inputs)
        dynamo_outputs = run(model, example_inputs, dynamo=True)
    for rid in range(len(eager_outputs)):
        cr = same(eager_outputs[rid], dynamo_outputs[rid])
        print(f"output id {rid}, correctness result {cr}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--device", choices=["cpu", "cuda"], default="cuda", help="specify device")
    args = parser.parse_args()
    DEVICE = args.device
    Model = load_model_by_name("fastNLP_Bert")
    m = Model(device=DEVICE, test="eval", jit=False)
    model, _ei = m.get_module()
    correctness_experiment(model, m.example_inputs)
```

Place this script in the torchbench root directory, name it `runx.py`, and run command: `python runx.py`, the output is:

```
output id 0, correctness result True
Accuracy failed for key name pred_start
output id 1, correctness result False
Accuracy failed for key name pred_start
output id 2, correctness result False
Accuracy failed for key name pred_start
output id 3, correctness result False
```

This script tests 4 different batches of inputs for both eager and torchdynamo+eager, the first batch returns the correct result, and the 2, 3, 4-th batches return incorrect results. It seems like, TorchDynamo still returns the old result (i.e., the output of the first batch) when given the rest of the batches as input.
The problem exists on both CUDA and CPU devices.

Please let me know if you could reproduce this problem, so that I know whether it is a bug in my environment, thanks!

**Comment by jansel** [comment]
Thanks for finding that, looks like a real issue to me.

**Comment by voznesenskym** [comment]
Starting to look at this finally, apologies for delay. I repro'd runx.py, thank you for providing it.

**Comment by voznesenskym** [comment]
Forgot to update this last night. I made a quick build with eval cache disabled. It passes the test (all True). Obviously much slower. My hypothesis of it caching the "forward" on the model causing issues is correct. I'll look at seeing if we can potentially capture inputs or other data on the frame (or caller frame even) to help make the lookup function correctly differentiate between different invocations. 



Cc @jansel

**Comment by voznesenskym** [comment]
Another update I forgot to add - 

This has nothing to do w/ eager, or at least, I can repro the same issue w/o eager.

Changing eager to a dummy fn:

```
def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
    return gm.forward
```

Causes the same problem.

I think we may need some new guard functions that properly capture args / inputs, or introduce some useful distinction other than the ones I see here:

___guarded_code, ___check_tensors, ___check_type_id, ___check_obj_id, ___is_grad_enabled, ___odict_getitem, ___tuple_iterator_len, ___tuple_iterator_getitem

**Comment by voznesenskym** [comment]
Probably a controversial PR, but this does make the test pass: https://github.com/pytorch/torchdynamo/pull/297

**Comment by voznesenskym** [comment]
A little more detail on the progress of the investigation.

The culprit causing the issue, w/ dynamo, is:

fastNLP/embeddings/bert_embedding.py:432

As this is the `forward`, that's a big culprit, so I had to read it a little to try to understand.

I narrowed the issue down to where we inspect the pool_method (in this case, 'first') to determine how to change the `outputs`.

The interesting line, specifically, is around:

`tmp = truncate_output_layer[_batch_indexes, batch_word_pieces_cum_length]`

Specifically, if we inspect the output_layer of the model between non-dynamo and dynamo, the `output_layer` in dynamo is always the same - the exact same one as the `output_layer` of the first non-dynamo run. If trace this back up to the source of the output_layer, bert_outputs, and sure enough, it follows the same pattern (identical across all of dynamo, matching the first non-dynamo, and then differing on the non-dynamo calls). I confirmed that the inputs and bert encoder (at least, in terms of sizes, layers, etc) were the same for all 8 runs. 

That means that, with dynamo, calls to:

```
bert_outputs, pooled_cls = self.encoder(word_pieces, token_type_ids=token_type_ids, attention_mask=attn_masks,
                                                output_all_encoded_layers=True)
```

produce the same exact result each time, given an identical input. 

Aside: What's a little odd to me, is that a model with `torch.manual_seed()` and `eval`, given identical input, *doesn't* produce the same result on subsequent inference passes.

Otherwise, the crux of the issue so far seems to be around the dynamo model producing the same exact result given the same exact input, most likely due to caching somewhere. I'm going to dig into the values on the bert encoder to see why they are (in non dynamo) / are not (in dynamo) changing as expected.

**Comment by voznesenskym** [comment]
So, the root cause is that the values behind `named_parameters` differ. This *should* make the nn module ineligible for caching, but it does not with today's guards, because we only check:

1) type_id match
2) for every key in `named_parameters`, ensure match 

However, if the forward pass of a model changes the weights/biases/values of the tensors within the layers... we do not detect that because following the above, (1) and (2) remained the same.

I updated https://github.com/pytorch/torchdynamo/pull/297/files to introduce another check, which we can move behind config if we decide to introduce it (its rather heavy, all things considered), this check is:

(3) For all values in the nn module's `named_parameters`, ensure a match.

**Comment by anijain2305** [comment]
@voznesenskym Do you know whats causing the state to change? I browsed through the code quickly, and also disabled the inlining of the following call, but it still fails. So, just wondering where is the state changing.

~~~
        bert_outputs, pooled_cls = self.encoder(word_pieces, token_type_ids=token_type_ids, attention_mask=attn_masks,
                                                output_all_encoded_layers=True)
~~~

**Comment by voznesenskym** [comment]
@anijain2305 are you asking why does the forward pass of the model cause it to change? 

If so, I am not sure. I hypothesized, at a distance, that it could be related to the dropouts. However, I turned the dropouts off (first by config, then by totally commenting them out), and it didn't fix it. 

Interestingly, calling `eval()` on either the nested BERT (aka self.encoder) or the outer model didn't make a difference. I don't know fastNLP_Bert well, at all, but I was thinking about digging in deeper into the layers and seeing where the forward pass might change something in the model state as a next step.

**Comment by anijain2305** [comment]
Yes, you answered my question.

Yes, it seems weird that the model state is changing in forward pass. From very quick look, nothing seemed suspicious.

**Comment by voznesenskym** [comment]
Thanks for taking a quick look, yea, its a weird one. My pr *does* fix it, but its not elegant, and checking modules for value comparison feels gross. Super open to ideas / suggestions.

**Comment by jansel** [comment]
If you enable [guard_nn_modules](https://github.com/pytorch/torchdynamo/blob/063bc13ee2e9f41e25e4b4ac96a62adcb944697b/torchdynamo/config.py#L39) does this fix the issue?

What is the actual value that is changing?  Based on the PR it seems like a tensor in the parameter list is the culprit.  Which parameter is it?

**Comment by voznesenskym** [comment]
> If you enable [guard_nn_modules](https://github.com/pytorch/torchdynamo/blob/063bc13ee2e9f41e25e4b4ac96a62adcb944697b/torchdynamo/config.py#L39) does this fix the issue?

No, it did not.

> What is the actual value that is changing? Based on the PR it seems like a tensor in the parameter list is the culprit. Which parameter is it?

That's what I am going to look into today and tomorrow. The parameter list is large, so I need to make a couple of utils that can dump changed parameters between invocations. I plan on using the non dynamo runs to understand what the delta is on inference pass +1, +2, etc, and then go from there.

**Comment by voznesenskym** [comment]
Okay, I have the test passing with a much simpler approach. 

I made a config driven option of `eval_model_named_parameter_checks_enabled` and `eval_model_named_parameter_checks`, which then drives `___check_obj_id` for the specified tensor.

Setting these to `eval_model_named_parameter_checks_enabled = True` (alongside `guard_nn_modules = True` ), and `eval_model_named_parameter_checks = ["weight"]` seems to be a pretty fast (I only see 11 new guard checks added here) solution.

I updated the attached PR, https://github.com/pytorch/torchdynamo/pull/297/files

**Comment by anijain2305** [comment]
Ok, I looked into this and seems to have stumbled upon something

https://github.com/fastnlp/fastNLP/blob/fb645d370f4cc1b00c7dbb16c1ff4542327c46e5/fastNLP/embeddings/bert_embedding.py#L479

![image](https://user-images.githubusercontent.com/13822661/171304516-8124f446-9820-40e7-bc8b-87c3ef4175d7.png)

Symptom - We ran the model twice with different inputs. But, TorchDynamo output did not change for the second input.

Cause - The line 479 mentioned above results in a `list` instead of a `Tensor` and TorchDynamo baked in the constant value of this `list` in the graph (confirmed by the following snippet of Fx graph) for line 482.

![image](https://user-images.githubusercontent.com/13822661/171304451-4cfeee5b-77e0-4a5c-8a68-b9da96d3520c.png)

Therefore, TorchDynamo frame ignores the list value for the second input, leading to incorrect results.

Coincidentally, I was working on an unrelated PR - https://github.com/pytorch/torchdynamo/pull/303 which adds more guards. The PR added CONSTANT_MATCH on the above list for this testcase. Therefore, it still bakes in the constant value of `list` but causes recompilation the next time a new `list` input comes.

Though the test passes with my PR, it is far from ideal, because we are essentially caching on the value of the list itself (equivalent to saying that we are caching on the value of the tensor). At the same time, this seems like to be badly written model (?, not sure).

cc @jansel

**Comment by voznesenskym** [comment]
@anijain2305  try my PR here: https://github.com/pytorch/torchdynamo/pull/297/files - it fixes it without caching on value. It adds an id check on module parameters.

**Comment by voznesenskym** [comment]
I was wrong here - only the value changes, the type id and object does not.

**Comment by voznesenskym** [comment]
Spoke to @anijain2305 about this, the leading theory was that there is an issue around caching of long_tensor, but actually, I think our trace makes that unclear by only showing the first few bits of the long tensor:

<img width="327" alt="image" src="https://user-images.githubusercontent.com/4755252/171517870-3b773cc5-0437-495c-a17c-e948f5369ef7.png">

Which makes it look like its the same

But actually, they differ at the end:

**Comment by voznesenskym** [comment]
Further update - it does not appear as though this is a named parameter issue - I ran a quick check, and in both the dynamo and non dynamo inference passes, the values of the named parameters of both models are the same.

**Comment by anijain2305** [comment]
tl;dr - The failure was due to a TorchDynamo bug - we were not adding constant guards for list of `numpy` integers.

TorchDynamo specializes on the list/tuple of integers, i.e., while creating Fx graphs, it bakes in the values inside list as constants into the graph. Simultaneously, it adds guards on those constant values of the list to recompile in case the list changes.

The bug revolves around handling of this list in TorchDynamo. There are two problems here with regards to this model, so lets untangle them
* Issue 1 - We can see that from the logs TorchDynamo did specialize on this list and baked it as a constant in captured Fx Graph. However, for some reason, the emitted guard was only for `LIST_LENGTH` and not `CONSTANT_MATCH`. This is the major clue.
* Issue 2 - In this model, the list is pretty large (~500 elements) and its actually as an input to `torch.LongTensor` operator. So, it raises the question of should we specialize on lists/tuples. Though important to discuss, this is a separate issue and is not a bug that causes failure for the concerned model. So, I will skip this one for now.

So, back to Issue 1 of why TorchDynamo did not emit `CONSTANT_MATCH` on list. This is because the list was not of type integers but numpy integers, and we missed that corner case. A much smaller repro is this

~~~
        @torchdynamo.disable
        def rand_gen():
            return list(np.array([random.randint(5, 10) for _ in range(10)]))

        def fn(x):
            random_list = rand_gen()
            z = torch.LongTensor(random_list)
            return x * z

        x = torch.ones(10) * 2

        random.seed(0)
        ref0 = fn(x)
        ref1 = fn(x)

        random.seed(0)
        with torchdynamo.optimize("eager"):
            res0 = fn(x)
            res1 = fn(x)

        self.assertTrue(same(ref0, res0))
        self.assertTrue(same(ref1, res1))
~~~

And the fix is in PR https://github.com/pytorch/torchdynamo/pull/313. Here, we simply allow numpy integers as constants while creating guards.

**Comment by voznesenskym** [comment]
Very nice! Great find. Confirmed it fixes this.

**Comment by jansel** [comment]
https://github.com/pytorch/torchdynamo/issues/70 is somewhat related.  Baking in int/float/numpy constants seems wrong.  We should turn them into graph inputs.

**Comment by anijain2305** [comment]
@xuzhao9 ping if there are more issues that you see.

**Comment by anijain2305** [comment]
Closing as the issues mentioned here are resolved.
